tests: Improve GitHub CI action

FlyGoat opened the pull request:
Improve up GitHub action by bumping software versions, use cache to preserve tool builds, and use Clang to speed up verilator.

FlyGoat committed:
ci: Speed up with better caching and clang, bump versions

Signed-off-by: Jiaxun Yang <jiaxun.yang@flygoat.com>

FlyGoat committed:
treewide: Fix "invalid escape sequence" warnings

Exposed with new tests.

Signed-off-by: Jiaxun Yang <jiaxun.yang@flygoat.com>

enjoy-digital commented:
Thanks a lot @FlyGoat. Would you mind having the `ci.yml` changes in another PR? I'm trying to synchronize changes for this between teh different repositories, so will maybe have more delay to look at it.

FlyGoat commented:
@enjoy-digital Done, would you be interested to test multiple Python versions with CI? I can implement it as well if you are planning to sync the change between repos.

Thanks

FlyGoat commented:
Hi @enjoy-digital,

Do you mind help me to merge those PRs so I can wire up those cores in CI?

https://github.com/litex-hub/pythondata-cpu-picorv32/pull/1
https://github.com/litex-hub/pythondata-cpu-mor1kx/pull/2
https://github.com/litex-hub/pythondata-cpu-cv32e40p/pull/3

Or maybe add me to litex-hub so I can help with maintaining pythondata packages :-)

enjoy-digital commented:
Hi @FlyGoat,

sorry, I just sent you an invite to be able to merge things directly. Feel free to merge these kind of PR directly. In case of doubt, feel free to ask, I'll try to be more reactive :)

FlyGoat commented:
@enjoy-digital Thanks a lot! I'll get those sorted.

FlyGoat committed:
test/test_integration: Test more CPU cores

Synthesis tested cpu list from litex.soc.cores.cpu to ensure
all CPUs are covered.

Refresh untested_cpu list to reflect latest status.

Signed-off-by: Jiaxun Yang <jiaxun.yang@flygoat.com>

FlyGoat committed:
ci: Test Python 3.8 and 3.12

Test Python 3.8 and 3.12 as our lower and upper bound of
supported python versions, since testing full python range
would require too much resources.

Signed-off-by: Jiaxun Yang <jiaxun.yang@flygoat.com>

FlyGoat committed:
ci: Build lm32 toolchain with crosstool-ng

Signed-off-by: Jiaxun Yang <jiaxun.yang@flygoat.com>

FlyGoat commented:
Hi @enjoy-digital,

This is ready for review. I'll look into individual failures after getting this merged.

Thanks
